### ISSUE TITLE
Start the flinkcluster is too slow.

### ISSUE THREAD

**Issue by Mrart** [GoogleCloudPlatform/flink-on-k8s-operator#190]
As per-job mode, I start a flink job on yarn, It only take 30s。 but the same jar running in k8s cluster , It takes 2mimute.  Has any good method to save start job times.

**Comment by functicons** [comment]
We need to first distinguish 1) the time used by the operator and 2) the time used the Flink itself. I haven't done any thorough analysis, but off the top of my head, I don't feel we can do much in the operator.

**Comment by Mrart** [comment]
I record it, and I will do it. share  experience.

**Comment by Mrart** [comment]
time waste like this.
` Normal  StatusUpdate  13m                FlinkOperator  Cluster status: Creating
  Normal  StatusUpdate  13m (x2 over 13m)  FlinkOperator  JobManager deployment status: NotReady
  Normal  StatusUpdate  13m (x2 over 13m)  FlinkOperator  ConfigMap status: Ready
  Normal  StatusUpdate  13m (x2 over 13m)  FlinkOperator  JobManager service status: Ready
  Normal  StatusUpdate  13m (x2 over 13m)  FlinkOperator  TaskManager deployment status: NotReady
  Normal  StatusUpdate  13m                FlinkOperator  JobManager deployment status changed: NotReady -> Ready
  Normal  StatusUpdate  12m                FlinkOperator  TaskManager deployment status changed: NotReady -> Ready
  Normal  StatusUpdate  12m                FlinkOperator  Cluster status changed: Creating -> Running
  Normal  StatusUpdate  12m (x2 over 12m)  FlinkOperator  Job status: Pending
  Normal  StatusUpdate  10m                FlinkOperator  (combined from similar events): Job status changed: Pending -> Running`

**Comment by Mrart** [comment]
@functicons , Can you give some suggest about this, why tm regest to jm need 1m.

**Comment by functicons** [comment]
I don't quite understand the timeline, could you rephrase it as something like:

<timestamp1> <event1>
<timestamp2> <event2>
...

**Comment by Mrart** [comment]
1. 41seconds after create flinkcluster ,the jobmanager status change to ready
2. 45seconds after the jobmanager ready, the taskmanager status change to ready
3. 118 seconds after the taskmanager status ready , the job status changet to running.

 "create flinkcluster"  -------(41s)-----"jobmanager ready"-------(45s)-----"taskmanager ready"----(118s) ------ "job running".  
some other things I user hadoop client in the init container to download jarfiles.

**Comment by functicons** [comment]
So the question is around why there is a 118s gap between TM ready and job running, right?

**Comment by Mrart** [comment]
No, TM ready and job ruuning gap is 118.

**Comment by functicons** [comment]
It should be that long, let me repro.

**Comment by Mrart** [comment]
@functicons   why set probe like this 
`var probe = corev1.Probe{
		Handler: corev1.Handler{
			TCPSocket: &corev1.TCPSocketAction{
				Port: intstr.FromInt(int(*jobManagerSpec.Ports.RPC)),
			},
		},
		TimeoutSeconds:      10,
		InitialDelaySeconds: 30,
		PeriodSeconds:       60,
		FailureThreshold:    5,
	}`

first liveness check time is 30s,and period is 60s?

**Comment by functicons** [comment]
Good point. I set these constants without careful thoughts about the performance, we can try decrease `InitialDelaySeconds` and `PeriodSeconds`, but increase `FailureThreshold` accordingly.

**Comment by Mrart** [comment]
@functicons  how long your initcontainer start in the job pods？ Have you record this？

**Comment by Mrart** [comment]
@functicons  as set TimeoutSeconds: 5, InitialDelaySeconds: 5, PeriodSeconds: 5, FailureThreshold: 30 OK? if that is ok, I will push it.

**Comment by Mrart** [comment]
Now we start the job pod after  the jm/tm pod sucess. Can start the job pod when the jm/tm pod start? before submit job check the jm start sucess?

**Comment by functicons** [comment]
If we start job submitter in parallel with JM/TM, and poll Flink API in it to make sure it is ready before submitting the job, that would be more complex for the operator. Because it's possible that JM/TM/Job status are ready, but Flink API server is not ready yet (might be stuck forever). In this case, it will show job Running, but actually not. I might be able to add a timeout in job submitter.

I'll verify the new constants in your previous message, I missed it  somehow.

**Comment by functicons** [comment]
In my test with https://github.com/GoogleCloudPlatform/flink-on-k8s-operator/pull/217, the gap between "cluster creating" and "job running is now around ~65s.

**Comment by functicons** [comment]
@Mrart Could you give it a test?

**Comment by Mrart** [comment]
@functicons  ～68s。I think we can start job pod and tm/jm together，and submit.sh check the fink ui is running. It means the initcontainer start with tm/jm together.

**Comment by functicons** [comment]
@Mrart  I'm wondering how critical it is to reduce the startup time? Because Flink jobs are typically long running streaming applications, startup time shouldn't matter that much. Let me know if I missed anything.

**Comment by Mrart** [comment]
We have build a platform , Different department in our company use this platform, It start slow ,let the user feel some thing not good. Or we will lost some users.
